File: event-bus-rabbit/src/main/java/pegasus/eventbus/rabbitmq/RabbitConnection.java
package pegasus.eventbus.rabbitmq;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

import org.apache.commons.lang.time.StopWatch;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.ShutdownListener;
import com.rabbitmq.client.ShutdownSignalException;

import pegasus.eventbus.amqp.AmqpConnectionParameters;

public class RabbitConnection implements ShutdownListener {

    private static final Logger          LOG                      = LoggerFactory.getLogger(RabbitConnection.class);
    private static final long            DEFAULT_RETRY_TIMEOUT    = 30000;

    private ConnectionFactory            connectionFactory        = new ConnectionFactory();
    private Connection                   connection;
    private long                         retryTimeout             = DEFAULT_RETRY_TIMEOUT;
    private Set<UnexpectedCloseListener> unexpectedCloseListeners = new HashSet<UnexpectedCloseListener>();

    private volatile boolean             isClosing;
    private volatile boolean             isInConnectionErrorState;

    public RabbitConnection(AmqpConnectionParameters connectionParameters) {
        connectionFactory.setUsername(connectionParameters.getUsername());
        connectionFactory.setPassword(connectionParameters.getPassword());
        connectionFactory.setVirtualHost(connectionParameters.getVHost());
        connectionFactory.setHost(connectionParameters.getHost());
        connectionFactory.setPort(connectionParameters.getPort());
    }

    public void open() throws IOException {
        if (connection == null || !connection.isOpen()) {
            isClosing = false;

            LOG.trace("Grabbing the connection instance from the factory.");

            connection = connectionFactory.newConnection();

            LOG.trace("Adding ShutdownListener to connection.");

            connection.addShutdownListener(this);
        }
    }

    public void setConnection(Connection connection) {
        this.connection = connection;
    }

    public boolean isOpen() {
        return connection != null ? connection.isOpen() : false;
    }

    public void close() throws IOException {
        if (connection.isOpen()) {
            isClosing = true;
            connection.close();
        }
    }

    public Channel createChannel() throws IOException {
        open();
        Channel channel = connection.createChannel();

        return channel;
    }

    public void shutdownCompleted(ShutdownSignalException signal) {
        if(isClosing){
            
        	LOG.info("Connection shutdown notice received as part of routine bus shutdown.  Taking no action.");
        	
        	return;
        } else {
	    	if (signal == null) {
	
	            LOG.info("Connection shutdown notice received unexpectedly.");
	
	        } else {
	
	            LOG.error("Connection shutdown exception received unexpectedly.", signal);
	        }
        }

        if (isInConnectionErrorState || isClosing) {
            return;
        }

        isInConnectionErrorState = true;

        StopWatch watch = new StopWatch();
        watch.start();
        try {

            while (watch.getTime() < retryTimeout) {
                
                LOG.info("Attempting to reopen connection.");
                
                try {
                    open();

                    LOG.info("Connection successfully reopened.");

                    isInConnectionErrorState = false;
                    break;
                } catch (Exception e) {

                    LOG.error("Attempt to reopen connection failed with error: " + e.getMessage(), e);

                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e1) {

                        LOG.warn("Attempt to reopen connection canceled because thread has been interrupted.");

                        // should we notify listeners in this case?
                        break;
                    }
                }
            }
            if (isInConnectionErrorState) {

                LOG.warn("Attempt to reopen connection permanently failed.");

            }

            notifyUnexpectedCloseListeners(!isInConnectionErrorState);

        } finally {
            watch.stop();
        }
    }

    public long getRetryTimeout() {
        return retryTimeout;
    }

    public void setRetryTimeout(long retryTimeout) {
        this.retryTimeout = retryTimeout;
    }

    public void attachUnexpectedCloseListener(UnexpectedCloseListener listener) {
        unexpectedCloseListeners.add(listener);
    }

    public void detachUnexpectedCloseListener(UnexpectedCloseListener listener) {
        unexpectedCloseListeners.remove(listener);
    }

    private void notifyUnexpectedCloseListeners(boolean successfullyReopened) {
        for (UnexpectedCloseListener listener : unexpectedCloseListeners) {
            listener.onUnexpectedClose(successfullyReopened);
        }
    }

    public interface UnexpectedCloseListener {

        void onUnexpectedClose(boolean successfulllyReopened);

    }

}

File: event-bus-rabbit/src/main/dotnet/rabbitmq/RabbitConnection.cs
using System;
using System.Diagnostics;
using System.Threading;

using log4net;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

using pegasus.eventbus.amqp;


namespace pegasus.eventbus.rabbitmq
{
	public class RabbitConnection
	{
        public event Action<bool> UnexpectedClose;

        private static readonly ILog LOG = LogManager.GetLogger(typeof(RabbitConnection));
        private static readonly TimeSpan DEFAULT_RETRY_TIMEOUT = new TimeSpan(0, 0, 30);

        private ConnectionFactory _connFactory;
        private IConnection _connection;
        private TimeSpan _retryTimeout = DEFAULT_RETRY_TIMEOUT;
        private bool _isClosing;


        public bool IsOpen
        {
            get
            {
                return (null != _connection) ? _connection.IsOpen : false;
            }
        }


		public RabbitConnection(AmqpConnectionParameters connectionParameters)
		{
            _isClosing = false;

            _connFactory = new ConnectionFactory();
            _connFactory.UserName = connectionParameters.Username;
            _connFactory.Password = connectionParameters.Password;
            _connFactory.VirtualHost = connectionParameters.VHost;
            _connFactory.HostName = connectionParameters.Host;
            _connFactory.Port = int.Parse(connectionParameters.Port);
		}


        public void Open()
        {
            if ((null == _connection) || (false == _connection.IsOpen))
            {
                _isClosing = false;
                _connection = _connFactory.CreateConnection();
                _connection.ConnectionShutdown += new ConnectionShutdownEventHandler(_connection_ConnectionShutdown);
            }
        }

        public void Close()
        {
            if ((null != _connection) && (_connection.IsOpen))
            {
                _isClosing = true;
                _connection.Close();
            }
        }

        public IModel CreateChannel()
        {
            this.Open();

            return _connection.CreateModel();
        }


        private void _connection_ConnectionShutdown(IConnection connection, ShutdownEventArgs reason)
        {
            // if we know we're closing, all is well
            if (_isClosing) return;

            if (null == reason)
            {
                LOG.Warn("Connection shutdown notice received unexpectedly");
            }
            else
            {
                LOG.WarnFormat("Connection shutdown exception received unexpectedly: [classId: {0}, initiator: {1}]",
                    reason.ClassId, reason.Initiator);
            }

            bool isInErrorState = true;

            Stopwatch watch = new Stopwatch();
            watch.Start();

            try
            {
                while (watch.Elapsed < _retryTimeout)
                {
                    try
                    {
                        this.Open();

                        LOG.Info("Connection successfully reopened");
                        isInErrorState = false;

                        break;
                    }
                    catch (Exception ex)
                    {
                        LOG.Warn("Reconnect attempt failed", ex);
                        Thread.Sleep(100);
                    }
                }

                if (isInErrorState)
                {
                    LOG.Error("Attempt to reopen rabbit connection permanently failed");
                }

                this.RaiseUnexpectedCloseEvent(!isInErrorState);
            }
            finally
            {
                watch.Stop();
            }
        }

        private void RaiseUnexpectedCloseEvent(bool successfullyReopened)
        {
            if (null != this.UnexpectedClose)
            {
                foreach (Action<bool> a in this.UnexpectedClose.GetInvocationList())
                {
                    try { a(successfullyReopened); }
                    catch (Exception ex) { LOG.Warn("Caught unhandled exception raising unexpected close event", ex); }
                }
            }
        }
	}
}

